sdk/data/azcosmos/cosmos_location_cache.go

```diff
@@ -0,0 +1,307 @@
+	locationUnavailabilityInfoMap     map[url.URL]locationUnavailabilityInfo
+	mapMutex                          sync.RWMutex
+	lastUpdateTime                    time.Time
+	enableMultipleWriteLocations      bool
+	unavailableLocationExpirationTime time.Duration
+}
+
+func newLocationCache(prefLocations []string, defaultEndpt url.URL) *locationCache {
+	return &locationCache{
+		defaultEndpt:                      defaultEndpt,
+		locationInfo:                      *newDbAcctLocationsInfo(prefLocations, defaultEndpt),
+		locationUnavailabilityInfoMap:     make(map[url.URL]locationUnavailabilityInfo),
+		unavailableLocationExpirationTime: defaultExpirationTime,
+	}
+}
+
+func (lc *locationCache) update(writeLocations []AcctRegion, readLocations []AcctRegion, prefList []string, enableMultipleWriteLocations bool) error {
+	lc.rwMutex.RLock()
+	nextLoc := copyDbAcctLocationsInfo(lc.locationInfo)
+	if prefList != nil {
+		nextLoc.prefLocations = prefList
+	}
+	lc.enableMultipleWriteLocations = enableMultipleWriteLocations
+	lc.refreshStaleEndpts()
+	if readLocations != nil {
+		availReadEndptsByLocation, availReadLocations, err := getEndptsByLocation(readLocations)
+		if err != nil {
+			lc.rwMutex.RUnlock()
+			return err
+		}
+		nextLoc.availReadEndptsByLocation = availReadEndptsByLocation
+		nextLoc.availReadLocations = availReadLocations
+	}
+
+	if writeLocations != nil {
+		availWriteEndptsByLocation, availWriteLocations, err := getEndptsByLocation(writeLocations)
+		if err != nil {
+			lc.rwMutex.RUnlock()
+			return err
+		}
+		nextLoc.availWriteEndptsByLocation = availWriteEndptsByLocation
+		nextLoc.availWriteLocations = availWriteLocations
+	}
+
+	nextLoc.writeEndpts = lc.getPrefAvailableEndpts(nextLoc.availWriteEndptsByLocation, nextLoc.availWriteLocations, write, lc.defaultEndpt)
+	nextLoc.readEndpts = lc.getPrefAvailableEndpts(nextLoc.availReadEndptsByLocation, nextLoc.availReadLocations, read, nextLoc.writeEndpts[0])
+	lc.lastUpdateTime = time.Now()
+	lc.rwMutex.RUnlock()
+	lc.rwMutex.Lock()
+	lc.locationInfo = nextLoc
+	lc.rwMutex.Unlock()
+	return nil
+}
+
+func (lc *locationCache) readEndpts() ([]url.URL, error) {
+	lc.mapMutex.RLock()
+	defer lc.mapMutex.RUnlock()
+	if time.Since(lc.lastUpdateTime) > lc.unavailableLocationExpirationTime && len(lc.locationUnavailabilityInfoMap) > 0 {
+		err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return lc.locationInfo.readEndpts, nil
+}
+
+func (lc *locationCache) writeEndpts() ([]url.URL, error) {
+	lc.mapMutex.RLock()
+	defer lc.mapMutex.RUnlock()
+	if time.Since(lc.lastUpdateTime) > lc.unavailableLocationExpirationTime && len(lc.locationUnavailabilityInfoMap) > 0 {
+		err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return lc.locationInfo.writeEndpts, nil
+}
+
+func (lc *locationCache) getLocation(endpoint url.URL) string {
+	firstLoc := ""
+	lc.rwMutex.RLock()
```

Comment:
Why do we need the lock here?

Reply:
This method is not using the map, so why do we need the lock on it?

Reply:
I understand methods working with `unavailableLocationExpirationTime` but what about those that don't?

Reply:
My concern is that this is the read path, yes, sure it can be concurrent, but so it is on other languages (C#, Java) and there are no locks there. If concurrency for these other parts (not related to the map) was a concern, then all the languages should have the locks.

Reply:
removed it